runtime/src/main/cpp/Boxing.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Memory.h"
+#include "Types.h"
+
+// C++ part of box caching.
+
+template<class T>
+struct KBox {
+  const ObjHeader header;
+  const T value;
+};
+
+// Keep naming of these in sync with codegen part.
+extern const KByte BYTE_RANGE_FROM;
+extern const KByte BYTE_RANGE_TO;
+
+extern const KChar CHAR_RANGE_FROM;
+extern const KChar CHAR_RANGE_TO;
+
+extern const KShort SHORT_RANGE_FROM;
+extern const KShort SHORT_RANGE_TO;
+
+extern const KInt INT_RANGE_FROM;
+extern const KInt INT_RANGE_TO;
+
+extern const KLong LONG_RANGE_FROM;
+extern const KLong LONG_RANGE_TO;
+
+extern const KBox<KByte> BYTE_CACHE[];
+extern const KBox<KChar> CHAR_CACHE[];
+extern const KBox<KShort> SHORT_CACHE[];
+extern const KBox<KInt> INT_CACHE[];
+extern const KBox<KLong> LONG_CACHE[];
+
+namespace {
+
+template<class T>
+bool isInRange(T value, T from, T to) {
+  return value >= from && value <= to;
+}
+
+
+template<class T>
+const ObjHeader *getCachedBox(T value, const KBox<T> cache[], T from) {
```

Comment:
All methods returning object references must be OBJ_GETTERs, even if they return permanent objects.